### PR TITLE
Feature/external adapters and callbacks

### PR DESCRIPTION
**Pull request by KensoDev**
So, this changes almost everything... so here's a quick explanation of what this is all about
1. The integration with the cache is further down the chain, it's when the cache is being written, so calls to `Rails.cache.write` and `Rails.cache.fetch` are also "catched" in cashier and can have `:tag` option
2. Refactored the class structure for the plugins.
3. I added the ability to add plugins to Cashier, it's still not 100% documented but it's fully working and tested
   Basically, you can add a plugin like an observer and receive callbacks when something happens in cashier.
   This is very useful when you want to add functionality like invalidate the URL based on the tag and so much more.
4. lots of changes to structured, added documentation on classes, many specs added.

Tests and coverage results
image

### CODE REVIEW

**Comment by ahawkins** [comment]
Do we really need the proc/lambda option?  I've never used it before. I'd really like to remove it actually because the code to get that module into ActionController is shoddy IMO. 

I think that going with this strategy from the beginning would've been better, especially since I know that all the options are just passed down. 

I generally like pull very much. The observers are very nice as well. Perhaps we should move it to use `ActiveSupport::Instrumentation` instead. That way we don't have to add anything directly to cashier, just move the instrumentation stuff into the abstract adapter. That would cut down on code complexity and still support all the uses cases. Plus, then you can keep track of the time it takes to expire each thing :D

I don't see a need for introducing the `Addons` namespace.

**Comment by KensoDev** [comment]
@Adman65 Ii never used proc as wel, if we don't need the procs, I can just remove the entire application_controller code and change the documentation.

About `ActiveSupport::Instrumentation` I just need to explore the path, I can refactor that for the next version.

I introduced the `Addons` namespace since we have `Adapters` and we have `Plugins` now. so I wanted it all to be under the same namespace.

I can put that to removal for the next version that will probably use the instrumentation.

**Comment by ahawkins** [comment]
@KensoDev If the tests pass without that module, then we should dump it.

I think that plugins can be it's own namespace because random crap could be added to there. 

I'm not sure if ActiveSupport::Instrumentation would solve your plugin problem. If you only need a way to notify external code, then it would do the trick. It depends on what kinda of information you need. 

You could do this something like this:

``` ruby
def expire(*tags)
  ActiveSupport::Instrumentation.instrument('expire.cashier', :data => tags) do
    adapter.expire tags
  end
end
```

Then your plugins could tie into that. Not sure if that would work, but it's an example.

**Comment by KensoDev** [comment]
@Adman65 tests would pass if you don't pass a proc into the `:tag` option.
I don't want to have procs other than in the controller code.

Regarding the `ActiveSupport` I will check.

**Comment by KensoDev** [comment]
@Adman65 Removed application controller code, no need for that, removed all traces and all tests pass of course.

**Comment by KensoDev** [comment]
@Adman65 Looked into supporting `ActiveSupport::Notifications.instrument` I don't really like it better then with plugins actually, I kinda like the plugins implementation just the way it is.

I think there's no added value with measuring the time it took.

perhaps in the future if cashier will grow further it will be useful.

Let's roll 0.5.0 without it and see what gives in the next releases

**Comment by ahawkins** [comment]
@KensoDev how are the plugins specifically better than `ActiveSupport::Notifications`?

**Comment by KensoDev** [comment]
@Adman65 not saying it's better, I said I like it better.

**Comment by ahawkins** [comment]
I want to merge this, but I am cautious about merging the plugin code as well. I need to review this a little more.

**Comment by KensoDev** [comment]
I can change the code to work with `ActiveSupport::Notifications` but I already have a plugin that's dependent on this version master branch.

Let me know if changing to `ActiveSupport::Notifications` will make merging easier, I think either way, this is a good thing to have in cashier.

**Comment by ahawkins** [comment]
I think that's better. Are you ok with doing that? If not, I can do it.

**Comment by KensoDev** [comment]
Since the plugin is not finished, I can do that, no problem.
I will have the pull request with the new code by tomorrow probably, lets sit on this till then

**Comment by ahawkins** [comment]
Please check out current master. I've made some major and minor revisions.

Major Revision: We can't integrate specifically with dalli because it's brittle. Instead, I've connected Cashier to the notifications coming out of ActiveSupport::Cache::Store.

Minor revisions: Moved adapter to Cashier.
